### ISSUE TITLE
Add Zap log context integration plugin

### ISSUE THREAD

**Issue by JustinBerger** [newrelic/go-agent#462]
## Summary

Currently there is a plugin for adjusting the formatting of logrus logs for ingestion into newrelic, but there isn't yet one for zap (which is supported for exporting logs from the agent).

## Desired Behaviour

Most important features would be setting the appropriate newrelic fields from the transaction context and changing the default field names in zap to match newrelic where possible.

## Possible Solution

https://github.com/JustinBerger/go-agent/tree/feature/nrzapplugin

## Additional context

Zap is already supported for getting logs out of the agent but doesn't have library support for getting correlated logs into newrelic.

**Comment by mpyw** [comment]
[Support for logging contextual metadata · Issue #654 · uber-go/zap](https://github.com/uber-go/zap/issues/654)

It would be quite a hard work since zap works without any contexts.

**Comment by iamemilio** [comment]
There are ways to work around this problem. Each golang logging framework has its own unique batch of problems that make it challenging to design an instrumentation package for them. We are going to look at zap shortly, and will let you know if we find it impractical to instrument. I'm sure there is some way forward though.

**Comment by mpyw** [comment]
Current my approach: **Copy a logger to register a hook just before calling a logger function**

```go
package log

import (
	"context"

	"github.com/newrelic/go-agent/v3/newrelic"
	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
)

type RootLogger struct {
	nr  *newrelic.Application
	zap *zap.Logger
}
func (root *RootLogger) Bg() ContextualLogger {
	return &contextualLogger{
		Logger: root.zap.WithOptions(newRecorderHook(root.nr)),
	}
}
func (root *RootLogger) Ctx(ctx context.Context) ContextualLogger {
	if txn := newrelic.FromContext(ctx); txn != nil {
		return &contextualLogger{
			Logger: root.zap.WithOptions(newRecorderHook(txn)),
		}
	}
	return root.Bg()
}

func newRecorderHook(recorder interface{ RecordLog(newrelic.LogData) }) zap.Option {
	return zap.Hooks(func(entry zapcore.Entry) error {
		recorder.RecordLog(newrelic.LogData{
			Timestamp: entry.Time.UnixMilli(),
			Severity:  entry.Level.String(),
			Message:   entry.Message,
		})
		return nil
	})
}

type ContextualLogger interface {
	Sugar() SugaredContextualLogger
	Named(name string) ContextualLogger
	Debug(msg string, fields ...zap.Field)
	Info(msg string, fields ...zap.Field)
	Warn(msg string, fields ...zap.Field)
	Error(msg string, fields ...zap.Field)
	DPanic(msg string, fields ...zap.Field)
	Panic(msg string, fields ...zap.Field)
	Fatal(msg string, fields ...zap.Field)
}
type SugaredContextualLogger interface {
	Named(name string) SugaredContextualLogger
	Debug(args ...any)
	Debugw(msg string, keysAndValues ...any)
	Debugf(template string, args ...any)
	Debugln(args ...any)
	Info(args ...any)
	Infow(msg string, keysAndValues ...any)
	Infof(template string, args ...any)
	Infoln(args ...any)
	Warn(args ...any)
	Warnw(msg string, keysAndValues ...any)
	Warnf(template string, args ...any)
	Warnln(args ...any)
	Error(args ...any)
	Errorw(msg string, keysAndValues ...any)
	Errorf(template string, args ...any)
	Errorln(args ...any)
	DPanic(args ...any)
	DPanicw(msg string, keysAndValues ...any)
	DPanicf(template string, args ...any)
	DPanicln(args ...any)
	Panic(args ...any)
	Panicw(msg string, keysAndValues ...any)
	Panicf(template string, args ...any)
	Panicln(args ...any)
	Fatal(args ...any)
	Fatalw(msg string, keysAndValues ...any)
	Fatalf(template string, args ...any)
	Fatalln(args ...any)
}

var _ ContextualLogger = (*contextualLogger)(nil)
var _ SugaredContextualLogger = (*sugaredContextualLogger)(nil)

type contextualLogger struct {
	*zap.Logger
}
type sugaredContextualLogger struct {
	*zap.SugaredLogger
}

func (logger *contextualLogger) Sugar() SugaredContextualLogger {
	return &sugaredContextualLogger{logger.Logger.Sugar()}
}
func (logger *contextualLogger) Named(name string) ContextualLogger {
	return &contextualLogger{logger.Logger.Named(name)}
}
func (logger *sugaredContextualLogger) Named(name string) SugaredContextualLogger {
	return &sugaredContextualLogger{logger.SugaredLogger.Named(name)}
}
```

Usage: 

```go
logger := &log.RootLogger{newrelicApp, zapLogger}

// for contextual logging:
logger.Ctx(req.Context()).Error("Error!", zap.Error(err))

// for background logging:
logger.Bg().Error("Error!", zap.Error(err))
```

It causes a shallow copy of `zap.Logger`, but since `zapcore.Core` is only copied as a pointer, it probably does not cause much slowdown.

**Comment by iamemilio** [comment]
This looks promising, but I will have to read into what data is available to hooks in zap to really know how viable this solution is. This isn't something that I'm going to have the time to do right now unfortunately.

**Comment by iamemilio** [comment]
This is an active roadmap item for us. I'm commenting to let people know we are working on a solution to this now!

**Comment by iamemilio** [comment]
@mpyw I reviewed your solution, and its great for collecting message, timestamp, and severity. However, similar to the issue we ran into with nrzerolog using the built in logger hooks, it would be difficult to capture the additional attributes users attach to their logs this way. This is something we eventually want to support, despite it being quite a headache in the go logging ecosystem. We are looking for a solution that can satisfy that requirement, if a natural solution exists. Thanks for submitting this though, that was very helpful, and we may fall back on it if we can't find a better one.
